File: day07/star14.py

class Directory:
    def __init__(self, name=None):
        self.name = name
        self.parent = None
        self.children = {}
        self.size = 0
        self.files = {}

def parse_line(line, pwd, directories):
    line = line.strip()
    if line.startswith('$ cd ..'):
        pwd = pwd.parent
    elif line.startswith('$ cd'):
        dirname = line[5:]
        fulldirname = f"{pwd.name}{dirname}/" if pwd else dirname
        try:
            pwd = directories[fulldirname]
        except KeyError:
            directories[fulldirname] = Directory(fulldirname)
            pwd = directories[fulldirname]
    elif line.startswith('$ ls'):
        pass
    elif line.startswith('dir'):
        dirname = line[4:]
        fulldirname = f"{pwd.name}{dirname}/" if pwd else dirname
        newdir = Directory(fulldirname)
        pwd.children[dirname] = newdir
        newdir.parent = pwd
        directories[fulldirname] = newdir
    else:
        size, name = line.split()
        size = int(size)
        pwd.files[name] = size
    return pwd, directories

def compute_size(directories, pwd=None):
    if pwd is None:
        pwd = directories['/']
    pwd.size += sum(pwd.files.values())
    for child in pwd.children.values():
        pwd.size += compute_size(directories, child)
    return pwd.size
    

with open('input.txt') as f:
    pwd = None
    directories = {}
    for line in f:
        pwd, directories = parse_line(line, pwd, directories)

compute_size(directories)

total_disk_size = 70000000
size_needed_for_update = 30000000
max_allowed = total_disk_size - size_needed_for_update

min_to_delete = directories['/'].size - max_allowed

smallest = directories['/'].size
for directory in directories.values():
    size = directory.size
    if size >= min_to_delete and size < smallest:
        smallest = size

print(f"{min_to_delete = }, {smallest = }")

File: day07/star14.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <algorithm>

std::vector<int> parse_lines(std::string&& fname) {
    std::string line;
    std::ifstream f (fname);
    std::vector<int> sizes {};
    std::vector<int> stack {};

    while (std::getline(f, line)) {
        if (line.substr(0, 7) == "$ cd ..") {
            stack.pop_back();
        } else if (line.substr(0, 4) == "$ cd") {
            stack.push_back(sizes.size());
            sizes.push_back(0);
        } else if (std::isdigit(line.at(0))) {
            int fsize;
            std::stringstream(line) >> fsize;
            for (int i=0; i<stack.size(); i++) {
                sizes[stack[i]] += fsize;
            }
        }
    }
    return sizes;
}

int find_infimum(std::vector<int> sizes, const int& threshold) {
    int smallest = *std::max_element(sizes.begin(), sizes.end());

    for (const auto& x : sizes) {
        if (x >= threshold && x < smallest) {
            smallest = x;
        }
    }
    return smallest;
}

int main() {
    std::vector<int> sizes = parse_lines("input.txt");

    int total_disk_size = 70000000;
    int size_needed_for_update = 30000000;
    int max_allowed = total_disk_size - size_needed_for_update;

    int min_to_delete = *std::max_element(sizes.begin(), sizes.end()) - max_allowed;
    std::cout << "min_to_delete = " << min_to_delete << std::endl;

    int sum = find_infimum(sizes, min_to_delete);
    std::cout << "smallest = " << sum << std::endl;
}